### PR TITLE
AP_Periph: added peripheral firmware

### PR DESCRIPTION
**Pull request by tridge**
This is a firmware CAN peripherals. Supported boards:

 - 128k flash UC4H board from jDrones (f103 based)
 - mRo CAN GPS
 - CUAV CAN GPS 

Many thanks to @pkocmoud and @jpkh for hardware and pinout assistance, and CUAV for hardware

To build for the f103 use:

 - ./waf configure --board f103-periph
 - ./waf AP_Periph

to build for the CUAV_GPS use:

 - ./waf configure --board CUAV_GPS
 - ./waf AP_Periph

to flash bootloader use:

 - st-flash write Tools/bootloaders/f103-periph_bl.bin 0x08000000
or
 - st-flash write Tools/bootloaders/CUAV_GPS_bl.bin 0x08000000

to flash main fw use:

 - st-flash write build/f103-periph/bin/AP_Periph.bin 0x8006000

or flash with the STM32 link utility on windows, at offset 0x8006000, or flash with uavcan_gui_tool fw update

milestones:
  - [x] flashes a LED
  - [x] UART writes
  - [x] can dynamic node allocation
  - [x] can node info response
  - [x] can bootloader and can fw update
  - [x] pin to hold to stay in bootloader (using PB6, the I2C1_SCL pin)
  - [x] remote I2C over CAN
  - [x] node restart
  - [ ] detect node ID conflicts
  - [ ] smarter default node ID
  - [ ] handle RTCM injection
  - [ ] support PPK via RXM RAW logging to FMU
  - [ ] add hardware can message filtering

### CODE REVIEW

**Comment by skyyuzhang** [comment]
I like it .hope it merged quickly

**Comment by skyyuzhang** [comment]
i am recently building my uavcan modules witch libcanard.  cause I think libuavcan seem to be a little hard . I have rearching zubax's projects. but it was too complicated for me . I am not expert by the way . hope your project can make clear and easy to read .so that might be able to help  a lot of people to learning libuavcan . thank you !!!!

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -169,6 +169,9 @@ def configure_env(self, cfg, env):
             '-Wfatal-errors',
             '-Wno-trigraphs',
         ]
+        env.CFLAGS += [
+            '-std=c99'
```
Comment (OXINARF) [inline Tools/ardupilotwaf/boards.py]: Any reason for not using c11 (or even gnu11)?

**Comment by pkocmoud** [comment]
Issues from testing as discussed with @tridge :

1) Add more RANDOM initial NODEID. Possible a HASH of the MCU serial number.
2) Add a NODEID collision detection mechanism and remediation scheme.
3) If a BLANK MCU is flashed with the bootloader only, then the Firmware is flashed over CAN, no automatic reboot occurs.
4) Add a mechanism to restart the automatic NODEID allocation routine. Including a confirmation period before committing to that ID.

File: Tools/ardupilotwaf/boards.py (hunk 1 above)
Comment (tridge) [reply Tools/ardupilotwaf/boards.py]: fixed

**Comment by tridge** [comment]
I believe this is ready for an initial merge. There are still more features I'd like, but they can come as a separate PR.

**Comment by pkocmoud** [comment]
@tridge Please merge. This has been great so far!

**Comment by tridge** [comment]
@pkocmoud yep, will do soon, I just want to give a chance for review by a few other devs now it is non-WIP

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp [hunk 2]
```diff
@@ -503,6 +509,7 @@ AP_GPS_UBLOX::read(void)
 // Private Methods /////////////////////////////////////////////////////////////
 void AP_GPS_UBLOX::log_mon_hw(void)
 {
+#ifndef HAL_NO_LOGGING
```
Comment (peterbarker) [inline libraries/AP_GPS/AP_GPS_UBLOX.cpp]: This double-negative is kind of painful.

File: libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp [hunk 3]
```diff
@@ -234,6 +233,7 @@ static void main_loop()
                                    pd.fault_thd_prio,
                                    pd.fault_icsr);
     }
+#endif
```
Comment (WickedShell) [inline libraries/AP_HAL_ChibiOS/HAL_ChibiOS_Class.cpp]: Comments for what this `#endif` is closing (IE `#endif // HAL_NO_LOGGING`) is a lot better for long term readability, particularly if there isn't a single line in the middle.

File: libraries/AP_Compass/AP_Compass.h [hunk 4]
```diff
@@ -124,6 +127,7 @@ friend class AP_Compass_Backend;
     // compass calibrator interface
     void cal_update();
 
+#if COMPASS_MOT_ENABLED
```
Comment (WickedShell) [inline libraries/AP_Compass/AP_Compass.h]: This triggers a crazy number of warnings when compiling:
```
In file included from ../../libraries/AP_AHRS/AP_AHRS.h:25,
                 from ../../libraries/AP_Avoidance/AP_Avoidance.cpp:6:
../../libraries/AP_Compass/AP_Compass.h:130:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
  130 | #if COMPASS_MOT_ENABLED
      |     ^~~~~~~~~~~~~~~~~~~
../../libraries/AP_Compass/AP_Compass.h:257:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
  257 | #if COMPASS_MOT_ENABLED
      |     ^~~~~~~~~~~~~~~~~~~
In file included from ../../libraries/AP_AHRS/AP_AHRS.h:25,
                 from ../../libraries/AP_Avoidance/AP_Avoidance.cpp:6:
../../libraries/AP_Compass/AP_Compass.h:356:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
  356 | #if COMPASS_CAL_ENABLED
      |     ^~~~~~~~~~~~~~~~~~~
../../libraries/AP_Compass/AP_Compass.h:477:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
  477 | #if COMPASS_CAL_ENABLED
      |     ^~~~~~~~~~~~~~~~~~~
../../libraries/AP_Compass/AP_Compass.h:481:5: warning: this use of "defined" may not be portable [-Wexpansion-to-defined]
  481 | #if COMPASS_MOT_ENABLED
      |     ^~~~~~~~~~~~~~~~~~~
```

File: libraries/AP_GPS/AP_GPS_UBLOX.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_GPS/AP_GPS_UBLOX.cpp]: it matches all the other HAL_NO_* defines we use when the default is for something to be enabled

File: libraries/AP_Compass/AP_Compass.h (hunk 4 above)
Comment (tridge) [reply libraries/AP_Compass/AP_Compass.h]: which compiler and target is this with? I don't get these warnings

Comment (WickedShell) [reply libraries/AP_Compass/AP_Compass.h]: STIL, gcc (GCC) 9.1.0

**Pull request merged**